File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     * [1,2,4,6]
     * prefix [1,1,2,8]
     * postfix [48,24,6,1]
     */
    productExceptSelf(nums) {
        const res = Array.from({length: nums.length}, () => 1);
        let prefix = 1
        for (let i=0; i<nums.length; i++) {
            res[i] = prefix
            prefix *= nums[i];
        }
        let postfix = 1;
        for (let i=nums.length-1; i>=0; i--) {
            res[i] *= postfix;
            postfix *= nums[i]
        }
        return res
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-3.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        res = []
        nums.sort()
        for i in range(len(nums)):
            if i and nums[i] == nums[i - 1]:
                continue
            l = i + 1
            r = len(nums) - 1
            while l < r:
                if nums[i] + nums[l] + nums[r] == 0:
                    res.append([nums[i], nums[r], nums[l]])
                    l += 1
                    r -= 1
                    while l < r and nums[l] == nums[l - 1]:
                        l += 1
                elif nums[i] + nums[l] + nums[r] < 0:
                    l += 1
                else:
                    r -= 1
        return res